src/Validation/Validator.php

```diff
@@ -95,50 +137,224 @@ private function validateParams(array $params, string $formId): array
 	}
 
 	/**
-	 * Validate files.
+	 * Validate files from the validation reference.
 	 *
 	 * @param array<string, mixed> $files Files to check.
-	 * @param array<string, mixed> $params Params for reference.
+	 * @param array<int|string, mixed> $validationReference Validation reference to check against.
 	 * @param string $formId Form Id.
 	 *
 	 * @return array<int|string, string>
 	 */
-	private function validateFiles(array $files, array $params, string $formId = ''): array
+	private function validateFiles(array $files, array $validationReference, string $formId = ''): array
 	{
 		$output = [];
 
+		// Check files.
 		foreach ($files as $fileKey => $fileValue) {
-			$input = $params[$fileKey] ?? [];
+			// Find validation reference by ID.
+			$reference = $validationReference[$fileKey] ?? [];
 
-			if (!$input) {
+			// Bailout if no validation is required.
+			if (!$reference) {
 				continue;
 			}
 
-			$fileName = $fileValue['name'] ?? '';
-			$fileSize = $fileValue['size'] ?? '';
-
-			$inputDetails = json_decode($input, true);
-			$inputData = $inputDetails['data'] ?? [];
-			$inputName = $inputDetails['name'] ?? '';
-
-			foreach ($inputData as $dataKey => $dataValue) {
-				switch ($dataKey) {
-					case 'validationAccept':
-						if (!empty($dataValue) && !$this->isFileTypeValid($fileName, $dataValue) && $inputData['validationRequired'] === '1') {
-							$output[$inputName] = sprintf($this->labels->getLabel('validationAccept', $formId), $dataValue);
+			// Loop all validations from the reference.
+			foreach ($reference as $dataKey => $dataValue) {
+				// Check validation for accept file types.
+				if ($dataKey === 'accept') {
+					foreach ($fileValue['name'] as $file) {
+						if (!empty($dataValue) && !$this->isFileTypeValid($file, $dataValue)) {
+							$output[$fileKey] = sprintf($this->labels->getLabel('validationAccept', $formId), $dataValue);
+							continue;
 						}
-						break;
-					case 'validationMinSize':
-						if (!empty($dataValue) && !$this->isFileMinSizeValid((int) $fileSize, (int) $dataValue * 1000) && $inputData['validationRequired'] === '1') {
-							$output[$inputName] = sprintf($this->labels->getLabel('validationMinSize', $formId), $dataValue);
+					}
+				}
+
+				// Check validation for size min/max.
+				foreach ($fileValue['size'] as $file) {
+					// Check validation for min size.
+					if ($dataKey === 'minSize') {
```

Comment:
This looks a bit odd. You're checking the key of the previous loop, inside another loop. 

Not sure I understand the flow here.

Reply:
yea the problem is in the the way you get the array of values in the multiple file. You will get something like this:
```php

[02-Nov-2021 13:34:54 UTC] Array
(
    [898b0eb2e944aece5e2ece26e5519d12] => Array
        (
            [name] => Array
                (
                    [0] => Image from iOS (1).jpg
                    [1] => Image from iOS (2).jpg
                    [2] => Image from iOS (3).jpg
                    [3] => Image from iOS (4).jpg
                )

            [type] => Array
                (
                    [0] => image/jpeg
                    [1] => image/jpeg
                    [2] => image/jpeg
                    [3] => image/jpeg
                )

            [tmp_name] => Array
                (
                    [0] => /private/var/tmp/phpIESHaY
                    [1] => /private/var/tmp/phpS2jJBm
                    [2] => /private/var/tmp/phpoCGnQi
                    [3] => /private/var/tmp/phpc2hX7F
                )

            [error] => Array
                (
                    [0] => 0
                    [1] => 0
                    [2] => 0
                    [3] => 0
                )

            [size] => Array
                (
                    [0] => 39637
                    [1] => 36483
                    [2] => 17234
                    [3] => 159145
                )

        )

)
```

So every key in the array reppresents one file this is why you need to check for the key in the parent. Stupid I know

Reply:
![butwait](https://user-images.githubusercontent.com/8638515/139862257-6ea01e45-d8d5-4a50-b52d-81404baea5b6.jpg)